compiler/rustc_lint/src/opaque_hidden_inferred_bound.rs

```diff
@@ -140,25 +140,51 @@ impl<'tcx> LateLintPass<'tcx> for OpaqueHiddenInferredBound {
                         }),
                         _ => None,
                     };
-                    cx.emit_spanned_lint(
-                        OPAQUE_HIDDEN_INFERRED_BOUND,
-                        pred_span,
-                        OpaqueHiddenInferredBoundLint {
-                            ty: cx.tcx.mk_opaque(
-                                def_id,
-                                ty::InternalSubsts::identity_for_item(cx.tcx, def_id),
-                            ),
-                            proj_ty: proj_term,
-                            assoc_pred_span,
-                            add_bound,
-                        },
-                    );
+                    let ty = cx
+                        .tcx
+                        .mk_opaque(def_id, ty::InternalSubsts::identity_for_item(cx.tcx, def_id));
+                    match opaque.origin {
+                        hir::OpaqueTyOrigin::FnReturn(_) | hir::OpaqueTyOrigin::AsyncFn(_) => {
+                            cx.emit_spanned_lint(
+                                OPAQUE_HIDDEN_INFERRED_BOUND,
+                                pred_span,
+                                OpaqueHiddenInferredBoundLint {
+                                    ty,
+                                    proj_ty: proj_term,
+                                    assoc_pred_span,
+                                    add_bound,
+                                },
+                            );
+                        }
+                        hir::OpaqueTyOrigin::TyAlias { .. } => {
+                            cx.tcx.sess.emit_err(OpaqueHiddenInferredBoundErr {
```

Comment:
hard errors in `rustc_lint` is somewhat :/

I also don't think we *should* error (or even lint, for that matter) here, tho

Reply:
opened a zulip thread about this https://rust-lang.zulipchat.com/#narrow/stream/315482-t-compiler.2Fetc.2Fopaque-types/topic/nested.20opaque.20bounds

Reply:
From a technical perspective, yeah, we shouldn't be doing hard errors inside of the lint infrastructure I think.